File: Geohexa.java
package geohexa;

public class Geohexa {
    
    private final static String BASE_36_DIGITS_OUTPUT = "0123456789abcdefghijkLmnopqrstuvwxyz";
    private final static String BASE_36_DIGITS = BASE_36_DIGITS_OUTPUT.toLowerCase();

    private boolean VERBOSE; // set to true to get detailed information on calculation steps

    void setVERBOSE () {
        VERBOSE = true;
    }

    public void cli(String[] args) {
        
        double lat = 0;
        double lon = 0;
        double acc = 3;
        String hexa = "";

        if (args.length == 2 || args.length == 3) {
            //System.out.println("Converting a lat and lon to a geohexa");
            
            try {
                lat = Double.parseDouble(args[0]);
                lon = Double.parseDouble(args[1]);
            } catch (NumberFormatException e) {
                System.err.println("\nLat and Lon must be decimal numbers e.g. 51.2 -0.125");
                System.exit(1);
            }
            
            if (lat < -90 || lat > 90 ) {
                System.err.println("\nLatitude must be in range -90 to 90");
                System.exit(1);
            }

            if (lon < -180 || lon > 180 ) {
                System.err.println("\nLongitude must be in range -180 to 180");
                System.exit(1);
            }
            
            if (args.length == 3) {
                try {
                    acc = Double.parseDouble(args[2]);
                } catch (NumberFormatException e) {
                    System.err.println("\nAccuracy must be a decimal number e.g. 1.5");
                    System.exit(1);
                }
            }

            hexa = latLonToGeohexa(lat, lon, acc);
            System.out.println("geohexa is " + hexa);
            Coordinate checkLatLon = geohexaToLatLon(hexa);
            System.out.println("Converted back to lat, lon: " + checkLatLon.getLat() + ", " + checkLatLon.getLon());
                
            return;
        }    

        if (args.length == 1) {
    
            //System.out.println("Converting a geohexa to a lat and lon");
            hexa = args[0].toLowerCase();
    
            if (! inRange(hexa)) {
                System.err.println("\nGeohexa digits must only be 0-9, a-z (or A-Z). Example: Qarj");
                System.exit(1);
            }
    
            Coordinate latLon = geohexaToLatLon(hexa);
            System.out.println("Lat: " + latLon.getLat() + " Lon: " + latLon.getLon());

            return;
        }

        System.out.println("Enter two or three arguments - a lat followed by a lon and optional accuracy");
        System.out.println("\nor\n\nEnter one argument - a geohexa");

    }
    
    public static boolean inRange(String hexa) {
        boolean inRange = true;
        for (char c : hexa.toCharArray()) {
            if (BASE_36_DIGITS.indexOf(c) == -1) {
                inRange = false;
            }
        }
        return inRange;
    }

    double distance(double lat1, double lon1, double lat2, double lon2) {

        double latDistance = Math.toRadians(lat2 - lat1);
        double lonDistance = Math.toRadians(lon2 - lon1);

        // Haversine formula
        double sinLat = Math.sin(latDistance / 2);
        double sinLon = Math.sin(lonDistance / 2);
        double a = sinLat * sinLat + Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2)) * sinLon * sinLon;
        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

        final int earthRadius = 6371;
        double distance = earthRadius * c * 1000; // convert to meters

        return distance;
    }

    private double prepareLat(double lat) {
        if (lat > 89.99999999) {
            lat = 89.99999999;
        }
        return lat + 90;
    }

    private double prepareLon(double lon) {
        if (lon == 180) {
            lon = -180; // 180E is same as 180W
        }
        return lon + 180;
    }

    // function only used for verbose output
    private double latBottom(String allLatgeo) {
        double lat = 0;

        if (allLatgeo.length() == 0) { // worst case for longitude size is at the equator, this also fits definition of null lat
            return lat;
        }

        double latHeight = 180;
        for (char d : allLatgeo.toLowerCase().toCharArray()) {
            lat = lat + (latHeight / 36) * BASE_36_DIGITS.indexOf(d);
            latHeight = latHeight / 36;
        }

        return lat - 90;
    }

    private Base36Remainder compress(double num, double cellSize) {
        int cell = (int) (num / cellSize);
        double remainder = num - ( (double) cell * cellSize);
        String base36 = Character.toString(BASE_36_DIGITS_OUTPUT.charAt(cell));

        if (VERBOSE) {
            System.out.println ("base36 digit: " + base36 + "(" + cell + ")  cell_size_degrees[" + cellSize + "]   fragment[" + num + "]  remainder_to_go[" + remainder + "]");
        }

        return new Base36Remainder(base36, remainder);
    }

    public String latLonToGeohexa(double lat, double lon, double... acc) {
        double accuracy = acc.length > 0 ? acc[0] : 3;
        if (accuracy < 0.00000001) { // need to be sensible about the accuracy to prevent division by zero
            accuracy = 0.00000001;
        }
        String geohexa = "";
        String allLatgeo = "";
        double latFragment = prepareLat(lat);
        double lonFragment = prepareLon(lon);
        double latUnit = 180;
        double lonUnit = 360;

        // null is a valid geohexa - represents 0, 0
        double error = distance(lat, lon, 0, 0);
        if (error < accuracy) {
            return "";
        }

        double latCellSize = 0;
        double lonCellSize = 0;
        if (VERBOSE) {
            latCellSize = distance(latBottom(allLatgeo), 0, latBottom(allLatgeo)+latUnit, 0);
        }

        Coordinate geoLatLon = new Coordinate();
        while (true) {
            lonUnit = lonUnit / 36;
            Base36Remainder lon36Remainder = compress(lonFragment, lonUnit);
            lonFragment = lon36Remainder.getRemainder();
            geohexa = geohexa + lon36Remainder.getBase36();

            geoLatLon = geohexaToLatLon(geohexa);
            error = distance(lat, lon, geoLatLon.getLat(), geoLatLon.getLon());

            if (VERBOSE) {
                lonCellSize = distance(latBottom(allLatgeo), 0, latBottom(allLatgeo), lonUnit);
                System.out.println("Current lat *lon* cell size meters: " + latCellSize + " , " + lonCellSize + "   error in meters[" + error + "]\n...");
            }

            if (error < accuracy) {
                break;
            }

            latUnit = latUnit / 36;
            Base36Remainder lat36Remainder = compress(latFragment, latUnit);
            latFragment = lat36Remainder.getRemainder();
            geohexa = geohexa + lat36Remainder.getBase36();
            allLatgeo = allLatgeo + lat36Remainder.getBase36();

            geoLatLon = geohexaToLatLon(geohexa);
            error = distance(lat, lon, geoLatLon.getLat(), geoLatLon.getLon());

            if (VERBOSE) {
                latCellSize = distance(latBottom(allLatgeo), 0, latBottom(allLatgeo)+latUnit, 0);
                System.out.println("Current *lat* lon cell size meters: " + latCellSize + " , " + lonCellSize + "   error in meters[" + error + "]\n\n");
            }

            if (error < accuracy) {
                break;
            }
        }

        return geohexa;
    }

    public Coordinate geohexaToLatLon(String hexa) {
        double lat = 0;
        double lon = 0;

        int i = 0;
        
        double latHeight = 180;
        double lonWidth = 360;

        for (char d : hexa.toLowerCase().toCharArray()) {
            i = i + 1;

            int remainder = Math.floorMod(i, 2);
            if (remainder == 1 ) { // we have a lon geohexa digit
                lon = lon + (lonWidth / 36) * BASE_36_DIGITS.indexOf(d);
                lonWidth = lonWidth / 36;
            } else {
                lat = lat + (latHeight / 36) * BASE_36_DIGITS.indexOf(d);
                latHeight = latHeight / 36;
            }
        }

        // calculate mid point of cell
        lat = lat - 90  + latHeight / 2;
        lon = lon - 180 + lonWidth / 2;

        if (VERBOSE) {
            System.out.println(hexa + " as lat, lon " + lat + " " + lon);
        }

        return new Coordinate(lat, lon);
    }

    public class Coordinate {
        private double mLat;
        private double mLon;
        public Coordinate () {
        }
        public Coordinate (double lat, double lon) {
            mLat = lat;
            mLon = lon;
        }
        public double getLat() {
            return mLat;
        }
        public double getLon() {
            return mLon;
        }
        public void setLat(double lat) {
            mLat = lat;
        }
        public void setLon(double lon) {
            mLon = lon;
        }
    }

    public final class Base36Remainder {
        private final String mBase36;
        private final double mRemainder;
        public Base36Remainder (String base36, double remainder) {
            mBase36 = base36;
            mRemainder = remainder;
        }
        public String getBase36() {
            return mBase36;
        }
        public double getRemainder() {
            return mRemainder;
        }
    }

}


/*


*/

/*
jrun geohexa.Run 51.481874 -0.112564
jrun geohexa.Run abcdefgh
ToDo:
 - Java variable naming conventions
 - unit tests for system out
 - lat_height instead of width
*/

File: geohexa.py
#!/usr/bin/env python3
version = "0.1.0"

import sys, argparse, math


def distance(lat1, lon1, lat2, lon2):

    latDistance = math.radians(lat2 - lat1)
    lonDistance = math.radians(lon2 - lon1)

    # Haversine formula
    sin_lat = math.sin(latDistance / 2)
    sin_lon = math.sin(lonDistance / 2)
    a = sin_lat * sin_lat + math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) * sin_lon * sin_lon
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))

    earth_radius = 6371
    distance = earth_radius * c * 1000
    # convert to meters

    return distance


def lat_valid(lat):
    if lat < -90 or lat > 90:
        return False
    return True


def lon_valid(lat):
    if lat < -180 or lat > 180:
        return False
    return True


def prepare_lat(lat):
    if lat > 89.99999999:
        lat = 89.99999999
    lat = lat + 90
    return lat


def prepare_lon(lon):
    if lon == 180:
        lon = -180  # 180E is same as 180W
    lon = lon + 180
    return lon


def compress(num, cell_size):
    cell, remainder = divmod(num, cell_size)
    base36 = base36digits_output[int(cell)]

    if args.verbose:
        print(
            "base36 digit: ",
            base36,
            "(" + str(int(cell)) + ")  cell_size_degrees[",
            cell_size,
            "]   fragment[",
            num,
            "]  remainder_to_go[",
            remainder,
            "]",
            sep="",
        )

    return base36, remainder


# function only used for verbose output
def lat_bottom(all_latgeo):
    lat = 0

    if len(all_latgeo) == 0:  # worst case for longitude size is at the equator, this also fits definition of null lat
        return lat

    lat_height = 180
    for d in all_latgeo.lower():
        lat = lat + (lat_height / 36) * base36digits.index(d)
        lat_height = lat_height / 36

    return lat - 90


def latlon_to_geohexa(lat, lon, accuracy=3):
    if accuracy < 0.00000001:  # need to be sensible about the accuracy to prevent division by zero
        accuracy = 0.00000001
    geohexa = ""
    all_latgeo = ""
    lat_fragment = prepare_lat(lat)
    lon_fragment = prepare_lon(lon)
    lat_unit = 180
    lon_unit = 360

    # null is a valid geohexa - represents 0, 0
    error = distance(lat, lon, 0, 0)
    if error < accuracy:
        return ""

    lat_cell_size = distance(lat_bottom(all_latgeo), 0, lat_bottom(all_latgeo) + lat_unit, 0)

    while True:
        lon_unit = lon_unit / 36
        longeo, lon_fragment = compress(lon_fragment, lon_unit)
        geohexa = geohexa + longeo

        geolat, geolon = geohexa_to_latlon(geohexa)
        error = distance(lat, lon, geolat, geolon)

        if args.verbose:
            lon_cell_size = distance(lat_bottom(all_latgeo), 0, lat_bottom(all_latgeo), lon_unit)
            print(
                "Current lat *lon* cell size meters: ",
                lat_cell_size,
                " , ",
                lon_cell_size,
                "   error in meters[",
                error,
                "]\n...",
                sep="",
            )

        if error < accuracy:
            break

        lat_unit = lat_unit / 36
        latgeo, lat_fragment = compress(lat_fragment, lat_unit)
        geohexa = geohexa + latgeo
        all_latgeo = all_latgeo + latgeo

        geolat, geolon = geohexa_to_latlon(geohexa)
        error = distance(lat, lon, geolat, geolon)

        if args.verbose:
            lat_cell_size = distance(lat_bottom(all_latgeo), 0, lat_bottom(all_latgeo) + lat_unit, 0)
            print(
                "Current *lat* lon cell size meters: ",
                lat_cell_size,
                " , ",
                lon_cell_size,
                "   error in meters[",
                error,
                "]\n\n",
                sep="",
            )

        if error < accuracy:
            break

    return geohexa


def geohexa_to_latlon(hexa):
    lat, lon = 0, 0
    i = 0
    lat_width = 180
    lon_width = 360

    for d in hexa.lower():
        i = i + 1

        integer, remainder = divmod(i, 2)
        if remainder == 1:  # we have a lon geohexa digit
            lon = lon + (lon_width / 36) * base36digits.index(d)
            lon_width = lon_width / 36
        else:
            lat = lat + (lat_width / 36) * base36digits.index(d)
            lat_width = lat_width / 36

    # calculate mid point of cell
    lat = lat - 90 + lat_width / 2
    lon = lon - 180 + lon_width / 2
    if args.verbose:
        print(hexa, "as lat, lon", lat, lon)
    return lat, lon


def in_range(hexa):
    in_range = True
    for c in hexa.lower():
        try:
            base36digits.index(c)
        except ValueError:
            in_range = False
    return in_range


parser = argparse.ArgumentParser(description="Convert a lat and lon to a geohexa, vice versa.")
parser.add_argument("--lat", dest="lat", required=False, action="store", help="Latitude")
parser.add_argument("--lon", dest="lon", required=False, action="store", help="Longitude")
parser.add_argument(
    "--acc",
    dest="acc",
    required=False,
    action="store",
    help="Accuracy for conversion to geohexa in meters, default is 3",
    default=3,
    type=float,
)
parser.add_argument("--hexa", dest="hexa", required=False, action="store", help="geohexa", default="")
parser.add_argument("--version", action="version", version=version)
parser.add_argument(
    "--verbose", action="store_true", dest="verbose", default=False, help="Will output extra info on calculation steps"
)
parser.add_argument(
    "--null", action="store_true", dest="null", default=False, help="Will output the lat and lon of a null geohexa"
)
args = parser.parse_args()

base36digits_output = "0123456789abcdefghijkLmnopqrstuvwxyz"
base36digits = base36digits_output.lower()

if args.lat and args.lon:
    lat = float(args.lat)
    lon = float(args.lon)
    if lat_valid(lat) and lon_valid(lon):
        geohexa = latlon_to_geohexa(lat, lon, accuracy=args.acc)
        print("geohexa is", geohexa)
        check_lat, check_lon = geohexa_to_latlon(geohexa)
        print("Converted back to lat, lon: ", check_lat, ", ", check_lon, sep="")
    else:
        print("Lat must be in range -90 to 90, Lon must be -180 to 180")
    sys.exit()

if args.hexa or args.null:
    if not in_range(args.hexa):
        print("geohexa digits must only be 0-9, a-z (or A-Z). Example: Qarj")
    else:
        lat, lon = geohexa_to_latlon(args.hexa)
        print("Lat:", lat, "Lon:", lon)
